### PR TITLE
[AMDGPU] Add iglp_opt(3) for simple mfma / exp interleaving

### PR DESCRIPTION
**Pull request by jrbyrnes**
Adds a minimal iglp_opt to do simple exp / mfma interleaving.

### CODE REVIEW

File: llvm/lib/Target/AMDGPU/AMDGPUIGroupLP.cpp [hunk 1]
```diff
@@ -832,7 +832,8 @@ void PipelineSolver::solve() {
 enum IGLPStrategyID : int {
   MFMASmallGemmOptID = 0,
   MFMASmallGemmSingleWaveOptID = 1,
-  MFMAExpInterleave = 2
+  MFMAExpInterleave = 2,
+  MFMAExpSimpleInterleaveID = 3
```
Comment (kerbowa) [inline llvm/lib/Target/AMDGPU/AMDGPUIGroupLP.cpp:836]: Do we document these strategies anywhere? Like what is the summary of the difference between `MFMAExpInterleave` and `MFMAExpSimpleInterleaveID`. Also it looks like ID suffix was forgotten for `MFMAExpInterleave` in a different patch?

File: llvm/lib/Target/AMDGPU/AMDGPUIGroupLP.cpp [hunk 2]
```diff
@@ -1845,6 +1846,50 @@ bool MFMAExpInterleaveOpt::applyIGLPStrategy(
   return true;
 }
 
+class MFMAExpSimpleInterleaveOpt final : public IGLPStrategy {
+private:
```
Comment (kerbowa) [inline llvm/lib/Target/AMDGPU/AMDGPUIGroupLP.cpp:1850]: NIT: remove `private:`

**Review by kerbowa — COMMENTED** [review]
Are the tests just missing check-lines?

File: llvm/lib/Target/AMDGPU/AMDGPUIGroupLP.cpp [hunk 3]
```diff
@@ -1845,6 +1846,50 @@ bool MFMAExpInterleaveOpt::applyIGLPStrategy(
...
+    DenseMap<int, SmallVector<SchedGroup, 4>> &SyncedSchedGroups,
+    AMDGPU::SchedulingPhase Phase) {
+  // Count the number of MFMA instructions.
+  unsigned MFMACount = 0;
+  for (const MachineInstr &I : *DAG)
+    if (TII->isMFMAorWMMA(I))
+      ++MFMACount;
+
+  const unsigned PipelineSyncID = 0;
+  SchedGroup *SG = nullptr;
```
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/AMDGPUIGroupLP.cpp:1879]: No sense in making this a free variable, just declare and initialize separate variables in the 2 individual uses

File: llvm/lib/Target/AMDGPU/AMDGPUIGroupLP.cpp [hunk 4]
```diff
@@ -1845,6 +1846,50 @@ bool MFMAExpInterleaveOpt::applyIGLPStrategy(
   return true;
 }
 
+class MFMAExpSimpleInterleaveOpt final : public IGLPStrategy {
+private:
+public:
+  bool applyIGLPStrategy(
+      DenseMap<int, SUnitsToCandidateSGsMap> &SyncedInstrs,
+      DenseMap<int, SmallVector<SchedGroup, 4>> &SyncedSchedGroups,
```
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/AMDGPUIGroupLP.cpp:1854]: What are these map keys? This could just be an array?

Comment (jrbyrnes) [reply llvm/lib/Target/AMDGPU/AMDGPUIGroupLP.cpp:1853]: The map keys are the SyncIDs for the SchedGroups. We try to enforce ordering of SchedGroups that have the same SyncIDs. If a user specifies SchedGroups with multiple SyncIDs, then we would construct multiple SchedGroup based pipelines that are overlayed on one another.

A multi-level structure works best with the Solver, but I suppose we could convert to a container for a vector of vectors with some SyncID -> index mapping. I think that should be done separately though.

**Review by kerbowa — APPROVED** [review]
LGTM

**Pull request merged**